bug(api): Update schema.rb

chrislessard opened the pull request:
This patch follows up on a recent PR to drop the rounds table from the schema.
It updates schema.rb (and makes some stylistic changes to the models) to reflect
this change.

freeatnet commented on app/javascript/app/components/courseForm.js:
```diff
@@ -58,24 +58,34 @@ class CourseForm extends React.Component {
+                                                  : 0
+                                            }
                                             min="0"
                                             onChange={eventKey =>
```
The argument should probably be called `event`. `eventKey` is a thing [from react-bootstrap](https://github.com/react-bootstrap/react-bootstrap/issues/432#issuecomment-77196712).

freeatnet commented on app/javascript/app/components/courseForm.js:
```diff
@@ -58,24 +58,34 @@ class CourseForm extends React.Component {
+                                                this.props.func.updateCourse(
                                                     course[0],
-                                                    eventKey
+                                                    eventKey,
```
Why pass the entire event to a data updater?

freeatnet commented on app/javascript/app/components/courseForm.js:
```diff
@@ -58,24 +58,34 @@ class CourseForm extends React.Component {
-                                                    eventKey
+                                                    eventKey,
+                                                    "hours",
+                                                    'positionHours'
```
It seems like we should use either the model's field name or the prop name only, and have a function that maps the one we provide to the other in `updateCourse`.

freeatnet commented on app/javascript/app/components/courseForm.js:
```diff
@@ -58,24 +58,34 @@ class CourseForm extends React.Component {
-                                                    eventKey
+                                                    event.target.value,
+                                                    "estimated_count",
+                                                    "estimatedPositions"
```
No luck creating a mapping?

michellemtchai replied:
I just added the mapping.

gabriellesc commented:
First thing tomorrow!

Le mer. 12 juil. 2017 à 19:42, Arseniy Ivanov <notifications@github.com> a
écrit :

> *@freeatnet* approved this pull request.
>
> Looks good to me, but I'd like a +1 from @gabriellesc
> <https://github.com/gabriellesc> on this.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/uoft-tapp/tapp/pull/80#pullrequestreview-49648706>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFTS5gWrRc_0hQIQ-FXYLQ2hs7JDgZxBks5sNVnlgaJpZM4OTLT8>
> .
>

gabriellesc commented on app/javascript/app/appState.js:
```diff
@@ -705,28 +705,26 @@ class AppState {
+        fetch.updateCourse(courseId, data, val, props);
+    }
+
+    getCourseAttributeMapping(props){
```
This function really shouldn't be in appState, given that it doesn't interact with the app state data in any way and isn't called from outside appState. I would put it directly in `updateCourse` as a nested function - or (although this is just a personal preference) I wouldn't make it a function at all, since it's only called once.

gabriellesc commented on app/javascript/app/appState.js:
```diff
@@ -737,13 +735,16 @@ class AppState {
 
     removeInstructor(courseId, index) {
```
We only want to generate a single Backbone 'change' event here, so that React only rerenders once!
Both `set` and `unset` will cause a change event, so we need to call `unset` as follows:
`unset(..., { silent: true })`
This tells Backbone not to generate a change event for the `unset`.

gabriellesc replied:
(1) Likewise, it looks like you are using `set` to change the instructors, and then calling `fetch.updateCourse`, which calls `appState.updateCourseAttribute` on success, which *itself* uses `set` to change the courses.
So there is one `set` call too many.

gabriellesc replied:
(2) We want to make sure that our data reflects the data in the database, so we only want to change our instructors list *if* we successfully changed the instructors list for this course in the database.

gabriellesc replied:
We can resolve both (1) and (2) by moving this `set` to a function called from the success callback of `fetch.updateCourse`.
However, it looks like (correct me if I'm wrong) this ends up being a duplicate of the [set](https://github.com/uoft-tapp/tapp/blob/master/app/javascript/app/appState.js#L878) in `appState.updateCourseAttribute`. So this first one is redundant and could be removed.

chrislessard committed:
bug(api): Update schema.rb

This patch follows up on a recent PR to drop the rounds table from the schema.
It updates schema.rb (and makes some stylistic changes to the models) to reflect
this change.
